### PR TITLE
Feature/add validators

### PR DESCRIPTION
**Pull request by cat-toreda**
added additional validators

### CODE REVIEW

File: src/validator/has-length-equal-to.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import {equalToFn} from './is-equal-to';
+
+export type STOpHasLengthEqualTo<CallerType> = (a: number) => CallerType;
+
+export const hasLengthEqualTo = (curr: any, target: number): boolean => {
+	if (curr.length <= -1) {
+		return false;
+	}
+
+	if (!curr.length) {
```
[the remark this answers is not in the document]
Comment (mbrich) [reply src/validator/has-length-equal-to.ts:15]: One more note, make sure you add a few unit tests which specifically test cases like comparing an empty array to 0.

File: src/validator/has-length-equal-to.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+import {equalToFn} from './is-equal-to';
+
+export type STOpHasLengthEqualTo<CallerType> = (a: number) => CallerType;
+
+export const hasLengthEqualTo = (curr: any, target: number): boolean => {
+	if (curr.length <= -1) {
```
Comment (mbrich) [inline src/validator/has-length-equal-to.ts:11]: This can be simplified a little to be:
```typescript
if (curr.length < 0) {
```
After changing the !curr.length check below, move this check below it. We always perform type tests prior to comparisons or value operations. The reason is simple: if a safety check is worth performing on a type to make sure it matches the expected type, then something bad can happen if the type is wrong. If the type is wrong, using comparisons, comparators, or any other operators can produce indeterminate behavior, so we haven no idea what it will do.

File: src/validator/has-length-greater-than-or-equal-to.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+import {greaterThanOrEqualToFn} from './is-greater-than-or-equal-to';
+
+export type STOpHasLengthGreaterThanOrEqualTo<CallerType> = (a: number) => CallerType;
+
+export const hasLengthGreaterThanOrEqualTo = (curr: any, target: number): boolean => {
+	if (curr.length <= -1) {
```
Comment (mbrich) [inline src/validator/has-length-greater-than-or-equal-to.ts:11]: - [ ] Similar to above, simplify statement to `if (curr.length < 0) {`
- [ ] Move length check below curr.length type check.

File: src/validator/has-length-greater-than-or-equal-to.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		return false;
+	}
+
+	if (!curr.length) {
```
Comment (mbrich) [inline src/validator/has-length-greater-than-or-equal-to.ts:15]: See above detailed comments elsewhere for more info, but update this to be a number type check.

File: src/validator/has-length-greater-than.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+import {greaterThanFn} from './is-greater-than';
+
+export type STOpHasLengthGreaterThan<CallerType> = (a: number) => CallerType;
+
+export const hasLengthGreaterThan = (curr: any, target: number): boolean => {
+	if (curr.length <= -1) {
```
Comment (mbrich) [inline src/validator/has-length-greater-than.ts:11]: - [ ] Similar to above, simplify statement to `if (curr.length < 0) {`
- [ ] Move length check below curr.length type check.

File: src/validator/has-length-greater-than.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+		return false;
+	}
+
+	if (!curr.length) {
```
Comment (mbrich) [inline src/validator/has-length-greater-than.ts:15]: - [ ] Same as elsewhere, update to be a number type check.

File: src/validator/has-length-less-than-or-equal-to.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+import {lessThanOrEqualToFn} from './is-less-than-or-equal-to';
+
+export type STOpHasLengthLessThanOrEqualTo<CallerType> = (a: number) => CallerType;
+
+export const hasLengthLessThanOrEqualTo = (curr: any, target: number): boolean => {
+	if (curr.length <= -1) {
```
Comment (mbrich) [inline src/validator/has-length-less-than-or-equal-to.ts:11]: - [ ] Similar to above, simplify statement to `if (curr.length < 0) {`
- [ ] Move length check below curr.length type check.

File: src/validator/has-length-less-than-or-equal-to.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+		return false;
+	}
+
+	if (!curr.length) {
```
Comment (mbrich) [inline src/validator/has-length-less-than-or-equal-to.ts:15]: Same as elsewhere, replace not operator with a number type check.

File: src/validator/has-length-less-than.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+import {lessThanFn} from './is-less-than';
+
+export type STOpHasLengthLessThan<CallerType> = (a: number) => CallerType;
+
+export const hasLengthLessThan = (curr: any, target: number): boolean => {
+	if (curr.length <= -1) {
```
Comment (mbrich) [inline src/validator/has-length-less-than.ts:11]: - [ ] Similar to above, simplify statement to `if (curr.length < 0) {`
- [ ] Move length check below curr.length type check.

File: src/validator/has-length-less-than.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+		return false;
+	}
+
+	if (!curr.length) {
```
Comment (mbrich) [inline src/validator/has-length-less-than.ts:15]: - [ ] Replace not operator with number type check

File: src/validator/is-ipv4-addr.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+export type STOpIsIpv4Addr<CallerType> = () => CallerType;
+
+export const isIpv4Addr = (curr: string): boolean => {
+	//Always a string.
+	if (typeof curr !== 'string') {
+		return false;
+	}
+	//Is valid if it has whitespace which can be trimmed with .trim()
+	//Invalid if it contains any other whitespace which cannot be trimmed with .trim()
+	if (!curr.trim()) {
```
Comment (mbrich) [inline src/validator/is-ipv4-addr.ts:16]: Using trim is good, however, there is both a performance cost to using it (tiny) and a consistency issue (more important). Anytime we're using string operations 2 or more times, it's a good idea to save the result to avoid performing the check over again. 

Imagine trim was instead a very expensive string search `O(n^3)` for some reason.
This might be something like:
```typescript
if (!curr.doExpensiveSearch('searchTerm')) {
   return false;
}

const myItem = curr.doExpensiveSearch('searchTerm');
```

We eat the expensive search once to see if what we're looking for exists. If not, we bail out it was performed only once. If it's found we then perform it again and eat the cost twice. Bringing it back to this function, we save the trimmed result and then use it in the second part.

```typescript
const trimmed = curr.trim();
if (!trimmed) {
   return false;
}

const pieces = trimmed.split('.');
```

The second issue was we validate the trimmed string initially, but then do not split the trimmed result, we split the original result.

File: src/validator/is-ipv4-addr.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+	//The first quad must be an integer between 1  and 255  (cannot be 0)
+	//All other quads must be an integer between 0 and 255
+	//Integers cannot have leading 0s
+	if (firstQuad <= 0 || firstQuad >= 256) {
+		return false;
+	}
+
+	if (
+		secondQuad <= -1 ||
```
Comment (mbrich) [inline src/validator/is-ipv4-addr.ts:39]: Good checks against the acceptable range, and good use of the early drop here when of the quads is invalid.
Given the block of comparisons, it's worthwhile to simplify slightly to make it a little easier to understand. 

- [ ] Change >= 256 checks to be > 255 and <= -1 checks to be < 0.

File: src/validator/is-ipv6-addr.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import {STRuleNodeType} from '../rule/node-type';
+
+export type STOpIsIpv6Addr<CallerType> = () => CallerType;
+
+export const isIpv6Addr = (current: string): boolean => {
+	if (typeof current != 'string') {
+		return false;
+	}
+
+	if (!current.trim()) {
```
Comment (mbrich) [inline src/validator/is-ipv6-addr.ts:14]: Similar to my ipv4 feedback, save the result of trim and do the truthy check against that

File: src/validator/is-ipv6-addr.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+		return false;
+	}
+
+	const section = current.split(':');
```
Comment (mbrich) [inline src/validator/is-ipv6-addr.ts:18]: Split the trimmed string, not original input string saved in the above feedback.

File: src/validator/is-port.ts [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import {isInteger} from './is-integer';
+
+export type STOpIsPort<CallerType> = () => CallerType;
+
+export const isPort = (curr: number): boolean => {
+	if (typeof curr !== 'number') {
+		return false;
+	}
+
+	if (curr <= -1) {
```
Comment (mbrich) [inline src/validator/is-port.ts:15]: When using `<=`, think about what you're trying to communicate in the code. Here, our valid range is `0` to `65343`, right? Therefore, there's a meaningful reason to check values strictly falling outside that range. When using less than or equal to, think about what you want to communicate and why you'd want to say less than or equal to vs strictly less than than. For instance, for loops iterating in reverse communicate something important by saying `i >= 0`. As long as `i` is a positive integer, we're okay. Conversely, saying `i < 0` is communicating that you're referring to all negative values for i. Here, because your range is specific, I'd simply do a strict less than / greater than to check values inside or outside the range.

- [ ] Simplify to `curr < 0` to make it a little easier to understand.

File: src/validator/is-port.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+		return false;
+	}
+
+	if (curr >= 65354) {
```
Comment (mbrich) [inline src/validator/is-port.ts:19]: - [ ] Change to be strictly greater than `65343` per feedback above.

**Review by mbrich — CHANGES_REQUESTED** [review]

**Review by mbrich — APPROVED** [review]

File: src/validator/has-property.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import {STRule} from '../rule/rule';
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+
+export type STOpHasProperty<CallerType> = (propName: string) => CallerType;
+
+function hasProperty(obj: any, propName: string): boolean {
```
Comment (mbrich) [inline src/validator/has-property.ts:9]: Checking whether a property exists using property name directly works but can produce false positives or negatives. It's better to use `obj.hasOwnProperty(propertyName)` to be sure, even though it is technically slower than direct index access. It checks whether the propertyName exists _directly_ on the object.

Note: It is a function on obj, so the call will throw if obj does not exist. Drop out if obj is not present.

- [ ] Replace hasProperty contents with obj.hasOwnProperty

File: src/validator/has-property.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+	if (typeof obj === 'undefined' || obj === null) {
+		return false;
+	}
+
+	const prop = obj[propName];
```
Comment (mbrich) [inline src/validator/has-property.ts:14]: This line:
```typescript
const prop = obj[propName];
```
becomes:
```typescript
const prop = obj.hasOwnProperty(propName);
```

**Review by mbrich — CHANGES_REQUESTED** [review]

File: src/validator/is-ipv6-addr.ts [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+	if (!section.length) {
+		return false;
+	}
+
+	if (section.length >= 9) {
+		return false;
+	}
+
+	if (section.length === 8 && !section.every(isValidSegment) && current.includes(doubleColon)) {
```
Comment (mbrich) [inline src/validator/is-ipv6-addr.ts:32]: Small performance adjustment here. Consider that `section.every` is an `O(n)` operation where `n` is the number of array elements. Here we're working with small numbers since the segment count is limited to 8, but between line 32 and line 36 we're doubling the number of operations needed by using every twice.

Move the every segment validation check above line 32 and use the resuult in both the `section.length` 8 and `section.length` 7 tests, and do an early drop and you can remove it from the length checks entirely.

Both checks bail out if all segments are not valid. And they use the same array input, so we can move it up and bail out if its false.
```typescript
const allValid = section.every(isValidSegment);
if (!allValid) {
   return false;
}


if (section.length === 8 && current.includes(doubleColon)) {
...
```

**Review by mbrich — APPROVED** [review]
